### PR TITLE
Bound missing-input recovery to 100 wallet ancestors

### PR DESCRIPTION
**Pull request by Kukks**
## Problem

When `sendrawtransaction` failed with “Missing inputs,” NBXplorer loaded every transaction for the tracked wallet, replayed every unconfirmed transaction, and retried the submitted transaction. That made one request capable of causing database work proportional to wallet history and RPC work proportional to the complete unconfirmed set.

The first bounded implementation narrowed recovery to direct parents only. Review identified two compatibility regressions in that approach: it could not recover a missing multi-generation chain, and repository result order could broadcast a child before its recovered parent.

## What changed

- Start from the submitted transaction's distinct direct input transaction IDs.
- Walk the tracked wallet's unconfirmed ancestor closure in bounded batches.
- Cap the complete discovered set—direct parents plus earlier ancestors—at 100 transaction IDs per request.
- Query only those transaction IDs for the supplied tracked source; unrelated wallet transactions are never loaded or replayed.
- Topologically order the recovered transactions so each known ancestor is broadcast before transactions that spend it.
- Retry the submitted transaction only when at least one recoverable ancestor was found.
- Preserve the original RPC error response when no ancestor is found or the final retry fails.
- Trigger recovery only for the established case-insensitive `Missing inputs` message prefix. A generic Bitcoin Core `RPC_VERIFY_ERROR` (`-25`) is no longer sufficient because it also represents non-repairable failures such as max-fee and internal mempool errors.

## Bounds and privacy

The database lookup set and rebroadcast set are both limited to at most 100 transaction IDs. Recovery cannot fall back to loading the wallet's full history, and it does not rebroadcast unrelated wallet activity. Deep or very wide chains beyond the cap keep the original broadcast failure instead of causing unbounded work.

## Compatibility

- Child-before-parent recovery remains supported.
- Multi-generation chains are recovered oldest-parent-first when their tracked ancestors fit within the 100-ID ceiling.
- Dependent direct parents are ordered correctly even if the database returns them in a different order.
- The legacy `Missing inputs` message match is retained for Bitcoin Core-compatible alternative nodes whose RPC error codes are not sufficiently specific.
- Normal broadcasts and `testmempoolaccept` behavior are unchanged.

## Validation

- Red/green focused tests cover generic `-25` rejection, case-insensitive missing-input detection, distinct-parent limits, dependent-parent ordering, and multi-generation ancestor traversal.
- `dotnet test NBXplorer.Tests/NBXplorer.Tests.csproj -c Release --filter FullyQualifiedName~BroadcastRecoveryTests --no-restore`: 4/4 passed.
- CI-equivalent `docker compose build` followed by `docker compose run --rm tests` from `NBXplorer.Tests`: 103/103 passed.
- `git diff --check`: passed.

## Scanner finding

Addresses `custos_buljdy`: the broadcast fallback can no longer amplify one malformed request into a replay of every unconfirmed wallet transaction.

## Rebase status

Rebased onto `master` after #564 merged. The temporary authorization-import commit was skipped because it is now in the base; this PR's effective diff contains only the bounded broadcast-recovery implementation and tests.

Post-rebase validation rebuilt the Docker test image with zero errors and passed all focused broadcast-recovery tests (4/4). GitHub CI is rerunning against rebased head `e21d2f5`.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Concept ACK

Findings
1. High: recovery fails for multi-generation unconfirmed chains  
NBXplorer/Controllers/MainController.cs:926-943  
For A -> B -> C, if the node has lost both A and B, broadcasting C loads only direct parent B. Broadcasting B fails because A is missing, the exception is swallowed, and retrying C also fails. The previous path replayed the wallet’s unconfirmed transactions in topological order. Recovery should traverse ancestors while retaining the 100-transaction bound.
2. Medium: direct parents are broadcast in unspecified order  
NBXplorer/Controllers/MainController.cs:927-943  
Repository results have no ordering. If C spends outputs from both A and B, while B also spends A, returning B before A causes B to fail permanently because each parent is tried only once. The selected transactions need topological ordering or failed parents need another bounded pass.
3. Medium: RPC_VERIFY_ERROR is not specific to missing inputs  
NBXplorer/Controllers/MainController.cs:39-41  
Bitcoin Core defines -25 as the generic transaction error and maps errors such as MAX_FEE_EXCEEDED and internal mempool errors to it, in addition to missing inputs. Such failures can now trigger up to 100 unrelated parent rebroadcasts and a retry. This reintroduces bounded privacy/RPC amplification for errors that parent recovery cannot fix.
CI passes, but the new tests do not exercise ancestor recovery, dependency ordering, or non-missing RPC_VERIFY_ERROR cases.

**Comment by NicolasDorier** [comment]
Concept ACK, I found three recovery edge cases that seem worth addressing:

1. **Multi-generation chains are no longer recoverable.** For an unconfirmed chain `A -> B -> C`, if the node has lost both `A` and `B`, broadcasting `C` only loads direct parent `B`. Broadcasting `B` fails because `A` is absent, that exception is swallowed, and retrying `C` fails too. The old path replayed the wallet transactions in topological order. Could the lookup walk a bounded ancestor closure (still capped at 100) instead?

2. **Even selected direct parents are broadcast in unspecified order.** Repository/SQL results are unordered. If `C` directly spends outputs from both `A` and `B`, while `B` also spends an output from `A`, returning `B` before `A` makes `B` fail; after `A` succeeds, `B` is not retried, so `C` still fails. The recovered set should be topologically ordered or retried in bounded passes. A test with dependent direct parents would cover this.

3. **`RPC_VERIFY_ERROR` is broader than missing inputs.** Bitcoin Core defines `-25` as the generic transaction error; `RPCErrorFromTransactionError` maps defaults including `MAX_FEE_EXCEEDED` and internal mempool errors to that same code. Those failures can therefore trigger parent rebroadcasts and a retry even though parents cannot repair them. This remains bounded, but it creates avoidable RPC/privacy activity. The error code alone does not identify the missing-input class. (Keep previous `Missing inputs` check, which work reliably with several shitcoins)

The scoped query and hard cap are good improvements, but I think the first two issues regress legitimate chained-transaction recovery.

**Comment by Kukks** [comment]
Addressed in `4272b20`.

- Recovery now walks the unconfirmed tracked-wallet ancestor closure, with the 100-ID cap applying to the entire discovered set.
- Recovered transactions are topologically ordered before broadcast, covering dependent direct parents and multi-generation chains.
- Recovery again requires the case-insensitive `Missing inputs` message prefix; generic `RPC_VERIFY_ERROR` failures no longer trigger wallet lookups or rebroadcasts.
- Added red/green tests for message classification, dependency ordering, and the `A -> B -> C` ancestor walk.

Fresh CI-equivalent validation after rebuilding and resetting the test stack: 103/103 tests passed.

File: NBXplorer/Controllers/MainController.cs [hunk 1]
```diff
@@ -904,22 +965,39 @@ public async Task<BroadcastResult> Broadcast(
 			{
 				rpcEx = ex;
 				Logs.Explorer.LogInformation($"{network.CryptoCode}: Transaction {tx.GetHash()} failed to broadcast (Code: {ex.RPCCode}, Message: {ex.RPCCodeMessage}, Details: {ex.Message} )");
-				if (trackedSourceContext.TrackedSource != null && ex.Message.StartsWith("Missing inputs", StringComparison.OrdinalIgnoreCase))
+				if (trackedSourceContext.TrackedSource != null && ShouldRecoverMissingInputs(ex.Message))
```
[the remark this answers is not in the document]
Comment (Kukks) [reply NBXplorer/Controllers/MainController.cs:968]: Not applying this additional code check. The maintainer explicitly requested retaining the previous `Missing inputs` message detection because several supported Bitcoin-compatible nodes do not reliably use Bitcoin Core's `RPC_VERIFY_ERROR` value for this case.

The privacy issue was the reverse condition: `RPC_VERIFY_ERROR` alone admitted unrelated failures such as max-fee errors. The current condition requires the established, case-insensitive `Missing inputs` prefix, so those generic `-25` failures no longer trigger a lookup. Requiring both signals would narrow valid recovery and regress the cross-node compatibility this review is restoring.

Covered by `MissingInputRecoveryRequiresTheMissingInputsMessage`, which accepts the message independent of RPC-code assumptions and rejects `Max fee exceeded`.

**Comment by Kukks** [comment]
Rebased onto current `master` after #564 merged. Git skipped the duplicate authorization-import commit, leaving only the three broadcast-recovery commits, headed by `e21d2f5`. Post-rebase Docker build completed with zero errors and the focused recovery suite passed 4/4. GitHub CI is rerunning on the rewritten head.
